### PR TITLE
Simple scheduler changes

### PR DESCRIPTION
**Pull request by rkdarst**
Hi,

This has two scheduler changes that should be non-controversial (first two from my other PR):
* Improve the random algorithm.  Now it's not stochastic and will always succeed on the first try, assuming parameters are valid.  Distribution is the same.
* Other fixes a bug I found while testing intervals.  I commented in another issue but there wasn't consensus there was a problem.  See the commit message and referenced commit that removed similar code: code hopefully makes it clear.

### CODE REVIEW

File: aware-core/src/main/java/com/aware/utils/Scheduler.java [hunk 1]
```diff
@@ -1238,27 +1239,47 @@ private boolean is_trigger(Schedule schedule) {
...
+            // an hour trigger and the schedule runner runs multiple times in that hour), then
+            // it will be set to false.  But, the most fine-grained condition takes precedence.
+            // For example, if we have an hour condition, it shouldn't run twice in an hour, but it
+            // should run multiple times in the same day.
+            Boolean execute_not_same_interval = true;
+
             Boolean execute_month = null;
             if (schedule.getMonths().length() > 0) {
-                execute_month = is_trigger_month(schedule);
+                if (is_trigger_month(schedule)) {
```
Comment (denzilferreira) [inline aware-core/src/main/java/com/aware/utils/Scheduler.java]: Missing the else to set the execute_month to false in our setup.

File: aware-core/src/main/java/com/aware/utils/Scheduler.java [hunk 2]
```diff
@@ -1238,27 +1239,47 @@ private boolean is_trigger(Schedule schedule) {
...
+                    execute_month = true;
+                    execute_not_same_interval = previous == null || !is_same_month(previous, now);
+                }
                 if (DEBUG) Log.d(Scheduler.TAG, "Trigger month: " + execute_month);
             }
 
             Boolean execute_weekdays = null;
             if (schedule.getWeekdays().length() > 0) {
-                execute_weekdays = is_trigger_weekday(schedule);
+                if (is_trigger_weekday(schedule)) {
```
Comment (denzilferreira) [inline aware-core/src/main/java/com/aware/utils/Scheduler.java]: Missing the else...

File: aware-core/src/main/java/com/aware/utils/Scheduler.java [hunk 3]
```diff
@@ -1238,27 +1239,47 @@ private boolean is_trigger(Schedule schedule) {
...
+                    execute_weekdays = true;
+                    execute_not_same_interval = previous == null || !is_same_weekday(previous, now);
+                }
                 if (DEBUG) Log.d(Scheduler.TAG, "Trigger weekday: " + execute_weekdays);
             }
 
             Boolean execute_hours = null;
             if (schedule.getHours().length() > 0) {
-                execute_hours = is_trigger_hour(schedule);
+                if (is_trigger_hour(schedule)) {
```
Comment (denzilferreira) [inline aware-core/src/main/java/com/aware/utils/Scheduler.java]: Missing the else...

File: aware-core/src/main/java/com/aware/utils/Scheduler.java [hunk 4]
```diff
@@ -1238,27 +1239,47 @@ private boolean is_trigger(Schedule schedule) {
...
+                    execute_hours = true;
+                    execute_not_same_interval = previous == null || !is_same_hour_day(previous, now);
+                }
                 if (DEBUG) Log.d(Scheduler.TAG, "Trigger hour: " + execute_hours);
             }
 
             Boolean execute_minutes = null;
             if (schedule.getMinutes().length() > 0) {
-                execute_minutes = is_trigger_minute(schedule);
+                if (is_trigger_minute(schedule)) {
```
Comment (denzilferreira) [inline aware-core/src/main/java/com/aware/utils/Scheduler.java]: Missing the else...

File: aware-core/src/main/java/com/aware/utils/Scheduler.java [hunk 5]
```diff
@@ -1572,31 +1594,21 @@ private void performAction(Schedule schedule) {
     public static ArrayList<Long> random_times(Calendar start, Calendar end, int amount, int interval_minutes) {
```
Comment (denzilferreira) [inline aware-core/src/main/java/com/aware/utils/Scheduler.java]: Looks good and efficient 👍

**Comment by rkdarst** [comment]
Updated by rebasing.  Good catches!

**Pull request merged**